aesd-char-driver/main.c

```diff
@@ -14,112 +11,309 @@
+        retval = -EINVAL;
+        return retval;
     }
-    else
+    data = filp->private_data;
+    if(count == 0)
     {
-    	dev->buffer_entry.buffptr = krealloc(dev->buffer_entry.buffptr, (dev->buffer_entry.size + count)*sizeof(char), GFP_KERNEL);
+        return retval;
     }
     
-    copy_from_user((void *)(&dev->buffer_entry.buffptr[dev->buffer_entry.size]), buf, count);
-    retval = count;
-    dev->buffer_entry.size += count;
-    for (int i=0; i<dev->buffer_entry.size; i++)
-    {
-    	if(dev->buffer_entry.buffptr[i] == '\n')
-    	{
-    		aesd_circular_buffer_add_entry(&dev->circular_buffer, &dev->buffer_entry);
-    		dev->buffer_entry.buffptr = NULL;
-    		dev->buffer_entry.size = 0;
-    	}
-    }
-    mutex_unlock(&aesd_device.m);
-    *f_pos = 0;
+    copy_buffer.buffer = kmalloc(count,GFP_KERNEL);
+    if(!copy_buffer.buffer)
+    {
+        retval = -ENOMEM;
+        kfree(copy_buffer.buffer);
+        return retval;
+    }
+    copy_buffer.size = count;
+    if (copy_from_user(copy_buffer.buffer, buf, count)) 
+    {
+	retval = -EFAULT;
+        kfree(copy_buffer.buffer);
+        return retval;
+    }
+
+    if (mutex_lock_interruptible(&data->lock))
+    {
+        retval = -EINTR;
+        kfree(copy_buffer.buffer);
+        return retval;
+    }
+    while(delim_loc != -1)
+    {
+        delim_loc = find_delimiter(&copy_buffer.buffer[start_ptr],(count - start_ptr));
+        mem_to_malloc = delim_loc==-1?(count - start_ptr):((delim_loc-start_ptr) +1);
+        if(allocate_memory(&data->working_buffer, mem_to_malloc)<0)
+        {
+            retval = -ENOMEM;
+	    mutex_unlock(&data->lock);
+            kfree(copy_buffer.buffer);
+            return retval;
+        }
+        memcpy((data->working_buffer.buffer + data->working_buffer.size),(copy_buffer.buffer+start_ptr),mem_to_malloc);
+        data->working_buffer.size += mem_to_malloc;
+        if(delim_loc != -1)
+        {
+            enqueue_buffer.buffptr = data->working_buffer.buffer;  
+            enqueue_buffer.size = data->working_buffer.size;
+            free_buffer = aesd_circular_buffer_add_entry(&data->circular_buffer,&enqueue_buffer);
+            if(free_buffer)
+            {
+                kfree(free_buffer);
+            }
+            data->working_buffer.size = 0;
+            start_ptr += delim_loc+1;
+        }
+        retval += mem_to_malloc;
+    }
+    mutex_unlock(&data->lock);
+    kfree(copy_buffer.buffer);
     return retval;
 }
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t ret_val;
+    struct aesd_dev *data;
+    if(!filp->private_data)
+    {
+        ret_val = -EINVAL;
+        return ret_val;
+    }
+    data = filp->private_data;
+    if (mutex_lock_interruptible(&data->lock))
+    {
+        ret_val = -EINTR;      
+        return ret_val;
+    }
+    ret_val = fixed_size_llseek(filp,off,whence, data->circular_buffer.size);
```

Comment:
@abaa3023 how could you modify this such that you wouldn't need to hold a lock during fixed_size_llseek?